src/contracts/core/DelegationManager.sol

```diff
@@ -528,131 +633,224 @@ contract DelegationManager is
+        for (uint256 i = 0; i < withdrawal.strategies.length;) {
+            require(
+                withdrawal.startTimestamp + strategyWithdrawalDelays[withdrawal.strategies[i]] <= block.timestamp,
+                "DelegationManager._completeReceiveAsTokens: withdrawalDelay period has not yet passed for this strategy"
+            );
 
-                _withdrawSharesAsTokens({
-                    staker: withdrawal.staker,
-                    withdrawer: msg.sender,
-                    strategy: withdrawal.strategies[i],
-                    shares: withdrawal.shares[i],
-                    token: tokens[i]
-                });
-                unchecked {
-                    ++i;
-                }
+            // Take already scaled staker shares and scale again according to current operator totalMagnitude
+            // This is because the totalMagnitude may have changed since withdrawal was queued and the staker shares
+            // are still susceptible to slashing
+            uint256 sharesToWithdraw =
+                _calculateSharesToCompleteWithdraw(withdrawal.scaledShares[i], totalMagnitudes[i]);
+
+            _withdrawSharesAsTokens({
+                staker: withdrawal.staker,
+                withdrawer: msg.sender,
+                strategy: withdrawal.strategies[i],
+                shares: sharesToWithdraw,
+                token: tokens[i]
+            });
+            unchecked {
+                ++i;
             }
-        } else {
-            // Award shares back in StrategyManager/EigenPodManager.
-            // If withdrawer is delegated, increase the shares delegated to the operator.
-            address currentOperator = delegatedTo[msg.sender];
-            for (uint256 i = 0; i < withdrawal.strategies.length;) {
-                require(
-                    withdrawal.startBlock + strategyWithdrawalDelayBlocks[withdrawal.strategies[i]] <= block.number,
-                    "DelegationManager._completeQueuedWithdrawal: withdrawalDelayBlocks period has not yet passed for this strategy"
-                );
+        }
+    }
+
+    /**
+     * @dev This function completes a queued withdrawal for a staker by receiving them back as shares
+     * in the StrategyManager/EigenPodManager. If the withdrawer is delegated to an operator, this will also
+     * increase the operator's scaled shares.
+     * This will apply any slashing that has occurred since the scaledShares were initially set in the Withdrawal struct
+     * by reading the original delegated operator's totalMagnitude at the time of withdrawal completion.
+     */
+    function _completeReceiveAsShares(
+        Withdrawal calldata withdrawal,
+        IERC20[] calldata tokens,
+        uint64[] memory totalMagnitudes
+    ) internal {
+        // read delegated operator for scaling and adding shares back if needed
+        address currentOperator = delegatedTo[msg.sender];
+
+        for (uint256 i = 0; i < withdrawal.strategies.length;) {
```

Comment:
a lot of logic is duplcated for checking the delay and looping through strategies between receive as shares